store/model.go

```diff
@@ -0,0 +1,276 @@
+		return ErrReadonlyTx
+	}
+	id := getEntityID(new)
+	if id == es.EmptyEntityID {
+		id = setNewEntityID(new)
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return errCantCreateExistingInstance
+	}
+
+	a := es.Action{
+		Type:       es.Create,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    new,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Save(updated interface{}) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+
+	id := getEntityID(updated)
+	key := t.model.dsKey.ChildString(id.String())
+	beforeBytes, err := t.model.datastore.Get(key)
+	if err == ds.ErrNotFound {
+		return errCantSaveNonExistentInstance
+	}
+	if err != nil {
+		return err
+	}
+
+	before := reflect.New(t.model.valueType.Elem()).Interface()
+	err = json.Unmarshal(beforeBytes, before)
+	if err != nil {
+		return err
+	}
+	a := es.Action{
+		Type:       es.Save,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   before,
+		Current:    updated,
+	}
+	t.actions = append(t.actions, a)
+
+	return nil
+}
+
+func (t *Txn) Delete(id es.EntityID) error {
+	if t.readonly {
+		return ErrReadonlyTx
+	}
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return ErrNotFound
+	}
+	a := es.Action{
+		Type:       es.Delete,
+		EntityID:   id,
+		EntityType: t.model.schema.Ref,
+		Previous:   nil,
+		Current:    nil,
+	}
+	t.actions = append(t.actions, a)
+	return nil
+}
+
+func (t *Txn) Has(id es.EntityID) (bool, error) {
+	key := t.model.dsKey.ChildString(id.String())
+	exists, err := t.model.datastore.Has(key)
+	if err != nil {
+		return false, err
+	}
+	return exists, nil
+}
+
+func (t *Txn) FindByID(id es.EntityID, v interface{}) error {
+	key := t.model.dsKey.ChildString(id.String())
+	bytes, err := t.model.datastore.Get(key)
+	if errors.Is(err, ds.ErrNotFound) {
+		return ErrNotFound
+	}
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(bytes, v)
+}
+
+func (t *Txn) Commit() error {
+	if t.discarded || t.commited {
+		return errAlreadyDiscardedCommitedTxn
+	}
+
+	events, err := t.model.eventcreator.Create(t.actions)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range events {
+		if err := t.model.dispatcher.Dispatch(e); err != nil {
+			return err // ToDo/Note: Important to document the implications of a partial dispatch
+		}
+	}
+	return nil
+}
+
+func (m *Model) Reduce(event es.Event) error {
+	// ToDo: distinguish local and remote events for proper locking
+	log.Debugf("reducer %s start", m.schema.Ref)
+	if event.Type() != m.schema.Ref {
+		log.Debugf("ignoring event from uninteresting type")
+		return nil
+	}
+
+	return m.eventcreator.Reduce(event, m.datastore, m.dsKey)
+}
+
+func (t *Txn) Discard() {
+	t.discarded = true
+}
+
+func getEntityID(t interface{}) es.EntityID {
```

Comment:
Some magic to extract the `EntityID` from an `interface{}` which represents an instance of a schema. Maybe be simplified; more important things pending. Works correctly.